{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### BI462 - Introduction to Bioinformatics\n",
    "# Lab 3: Short Read Assembly using `velvet`\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this lab you will learn to practice the popular short-read assembler, `velvet`, a De-Bruijn Graph based assembler on the single-end short reads generated by Illumina sequencer. \n",
    "\n",
    "The data for assembled is from **Staphylococcus aureus (金黄色葡萄球菌) USA300** which has a genome of around **3MB **. The reads are Illumina single-end library."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Download and compile the assembler\n",
    "\n",
    "```bash\n",
    "mkdir -p ~/bioinformatics/velvet\n",
    "cd ~/bioinformatics/velvet\n",
    "wget http://www.ebi.ac.uk/~zerbino/velvet/velvet_1.2.03.tgz\n",
    "tar xzf velvet_1.2.03.tgz; cd velvet_1.2.03\n",
    "make velveth velvetg\n",
    "```\n",
    "These commands will generate a few executable files:\n",
    "- `velveth`: generates hash information.\n",
    "- `velvetg`: conducts the de-Bruijn graph assembling into contigs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Download the data\n",
    "\n",
    "The reads can be obtained from the **Sequence Read Archive (SRA)**. For the following example use the run data **SRR022825** and **SRR022823** from the SRA Sample **SRS004748**.\n",
    "\n",
    "```bash\n",
    "cd ~/bioinformatics/velvet\n",
    "mkdir data\n",
    "wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR022/SRR022825/SRR022825.fastq.gz\n",
    "wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR022/SRR022823/SRR022823.fastq.gz\n",
    "```\n",
    "\n",
    "Now you are ready to process the compressed FASTQ files with velvet."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. `velveth`\n",
    "\n",
    "`velveth` supports the following input file formats and read categories:\n",
    "\n",
    "### 3.1 Input file formats\n",
    "- **FASTA** (default) or **FASTA.gz**\n",
    "- **FASTQ** or **FASTQ.gz**\n",
    "- **SAM** or **BAM**\n",
    "- **Eland**\n",
    "- **Gerald**\n",
    "\n",
    "### 3.2 Read categories\n",
    "- **short**: short reads\n",
    "- **shortPaired**: short paired-end reads\n",
    "- **short2**: same as short, but for a separate insert-size library\n",
    "- **shortPaired2**: short paired-end reads in two separate files\n",
    "- **long**: long reads\n",
    "\n",
    "Now run `velveth` for the reads in **SRR022825.fastq.gz** and **SRR022823.fastq.gz** using the following options:\n",
    "- A De Bruijn graph with **k-mer of 25**\n",
    "- An output directory called **kmer_25**\n",
    "\n",
    "```bash\n",
    "velveth kmer_25 25 -fastq.gz -short data/SRR022825.fastq.gz data/SRR022823.fastq.gz\n",
    "```\n",
    "\n",
    "This will take a short while and ends with some files in the output directory."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=\"red\">$\\S$ Exercise 1</font>\n",
    "\n",
    "1. What did you find in the folder kmer_25?\n",
    "2. Describe the content of the two velveth output files.\n",
    "3. What does the **Log** file store for you?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. `velvetg`\n",
    "\n",
    "Now run `velvetg` on your output directory, with the commands:\n",
    "```bash\n",
    "time velvetg run_25\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=\"red\">$\\S$ Exercise 2</font>\n",
    "\n",
    "1. What extra files do you see in the folder kmer_25?\n",
    "2. What do you suppose they might represent?\n",
    "3. In the **Log** file in kmer_25, what is the **N50**?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### $\\S$ What is N50?\n",
    "\n",
    "Broadly speaking, N50 is the median (not average) of a sorted data set using the length of a set of sequences. Usually it is the length of the contig whose length, when added to the length of all longer contigs, makes a total greater that half the sum of the lengths of all contigs. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=\"red\">$\\S$ Exercise 3</font>\n",
    "\n",
    "We can use a java program to calculate the truncated N25, N50, and N75:\n",
    "```bash\n",
    "java -jar contrib/gnx.jar -min 100 -nx 25,50,75 kmer_25/contigs.fa\n",
    "```\n",
    "\n",
    "Does the value of N50 agree with the value stored in the Log file? If not, why do you think this might be?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Improving the assembly using some options\n",
    "\n",
    "In order to improve our results, take a closer look at the standard options of velvetg by typing `velvetg` without any parameters."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.1 `-cov_cutoff`\n",
    "\n",
    "Clearly **-cov_cutoff** will allow you to exclude contigs with low-coverage kmer coverage, implying unacceptably poor quality. \n",
    "\n",
    "### 5.2 `-exp_cov`\n",
    "\n",
    "The **-exp_cov** switch is used to give `velvetg` a coverage to expect. If the expected coverage of any contig is substantially in excess of the suggested expected value, maybe this would indicate a repeat. \n",
    "\n",
    "We can use a **contributed software**, written in Perl, to **visualize the distribution** and **estimate the expected coverage of $k$-mers**.\n",
    "```bash\n",
    "contrib/estimate-exp_cov/velvet-estimate-exp_cov.pl kmer_25/stats.txt\n",
    "```\n",
    "\n",
    "This will generate a **weighted ASCII histogram**, with which you can determine the expected coverage and figure out a possible coverage threshold."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color='red'>$\\S$ Exercise 4</font>\n",
    "\n",
    "1. What is the possible expected coverage of $k$-mers?\n",
    "2. What is the coverage cutoff to remove the likely noise,  in your opinion?  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.3 Rerun `velvetg` using some options\n",
    "\n",
    "```bash\n",
    "# backup the previous assembled contigs\n",
    "cp kmer_25/contigs.fa kmer_25/contigs.fa.0\n",
    "\n",
    "# run with coverage cutoff of 6\n",
    "time velvetg kmer_25 -cov_cutoff 6\n",
    "cp kmer_25/contigs.fa kmer_25/contigs.fa.1\n",
    "\n",
    "time velvetg run_25 -exp_cov 14\n",
    "cp kmer_25/contigs.fa kmer_25/contigs.fa.2\n",
    "time velvetg run_25 -cov_cutoff 6 -exp_cov 14\n",
    "cp kmer_25/contigs.fa kmer_25/contigs.fa.3\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=\"red\">$\\S$ Exercise 5</font>\n",
    "\n",
    "1. What is the N50 with no parameter?\n",
    "2. What is the N50 with `-cov_cutoff 6`?\n",
    "3. What is the N50 with `-exp_cov 13`?\n",
    "4. What is the N50 with `-cov_cutoff 6 -exp_cov 13`?\n",
    "5. Did you notice a trend in the time `velvetg` took to run? If so, can you explain why that might be?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now you can try using different values for both `-cov_cutoff` and `-exp_cov`. Similarly, you can explore the further use of the other options:\n",
    "- `-max_coverage`: cut-off value for the upper range (like `cov_cutoff` for the lower range).\n",
    "- `-max_branch_length`: length of branch to look for bubble.\n",
    "- `-unused_reads`: write unused reads into file.\n",
    "- `-amos_file`: write AMOS message file.\n",
    "- `-read_trkg`: tracking read (more memory usage) - automatically on for certain operations.\n",
    "- etc."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=\"red\">$\\S$ Exercise 6</font>\n",
    "\n",
    "If you want to explore the behaviour of velvet even further, you could experiment with the following.\n",
    "1. Reduce the sequence coverage by only choosing one input file for the assembly e.g. SRR022825.fastq.gz.\n",
    "2. Increase the sequence coverage by downloading further files from the same sample SRS004748.\n",
    "\n",
    "How did the N50 change?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## 6. Visualization of the contigs\n",
    "\n",
    "In particular, look at the **-amos_file** parameter which instructs velvetg to create a version of the\n",
    "assembly that can be viewed with a program called **tablet**, written in java.\n",
    "```bash\n",
    "velvetg kmer_25 -cov_cutoff 6 -exp_cov 13 -amos_file yes\n",
    "```\n",
    "This, will generate a file named `velvet_asm.afg`, which can be visualized using `tablet`.\n",
    "\n",
    "You can download the reference genome sequence to assist the viewing."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
